### PR TITLE
FindCreatureWithOptions with StringId option cherrypicked from Master

### PR DESCRIPTION
**Pull request by ghost**
Add the ability to search for npc with StringId to inprove scripting.  Will look into adding support for Gameobjects, conditions, and SAI in the future if this is added.

Picked from:
Core/Entities: Added possibility to inherit StringIds from other entities #29908
Core/Creatures: Fix logic when searching for nearby creatures by StringId #28669
[Creature/Scripting]: Implement StringId for Creatures, assigning a… #28500
Core/Objects: Implement FindNearestCreatureWithOptions helper function #28488

**Changes proposed:**

-  Add FindNearestCreatureWithOptions and StringId to 335.

**Tests performed:**

Compiles and tested in game.

**Known issues and TODO list:** (add/remove lines as needed)

- [x]  Fix issue that is making it crash.
- [x]  Fix issue only working on closest npc without the search StringId.

### CODE REVIEW

**Comment by malcrom** [comment]
image

**Comment by Shauren** [comment]
m_stringIds is missing initialization in constructor

**Comment by malcrom** [comment]
No longer crashing but only working if creature doesn't have string id.
image
Worked on Kira as she passed by...

**Comment by Odyssey** [comment]
> No longer crashing but only working if creature doesn't have string id.

You are missing https://github.com/TrinityCore/TrinityCore/commit/795edfaf98e065b8eaa76683e0cd1a4211b0847d

**Comment by malcrom** [comment]
> > No longer crashing but only working if creature doesn't have string id.
> 
> You are missing [795edfa](https://github.com/TrinityCore/TrinityCore/commit/795edfaf98e065b8eaa76683e0cd1a4211b0847d)

Ty it must have be updated after.

**Comment by malcrom** [comment]
Working :)

File: sql/updates/world/3.3.5/xxxx_xx_xx_xx.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+DELETE FROM `trinity_string` WHERE `entry` = 5089;
+INSERT INTO `trinity_string` (`entry`,`content_default`) VALUES
+(5089, 'Template StringID: %.*s\r\nSpawn StringID: %.*s\r\nScript StringID: %.*s');
+
+ALTER TABLE `creature` ADD `StringId` varchar(64) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL AFTER ScriptName;
```
Comment (CraftedRO) [inline sql/updates/world/3.3.5/xxxx_xx_xx_xx.sql:5]: `ScriptName` backticks if needed

Comment (malcrom) [reply sql/updates/world/3.3.5/xxxx_xx_xx_xx.sql:5]: You guys give me enough information to say "What?"

Comment (malcrom) [reply sql/updates/world/3.3.5/xxxx_xx_xx_xx.sql]: Thats what I get for copying someone elses SQL.

**Comment by Shauren** [comment]
Closing this one - there are missing pieces from original implementation plus its a manual port instead of using `git cherry-pick`

**Comment by CraftedRO** [comment]
malcrom allow me to make a joke in your PR, in the future if we have to reconsider our project name, that might be CherrypickCore 😆

**Comment by malcrom** [comment]
> malcrom allow me to make a joke in your PR, in the future if we have to reconsider our project name, that might be CherrypickCore 😆

Al least I got it added. I can't learn all the stuff they want you to do here. This is not my profession.

**Pull request closed without merging**